Unify file encoding when a cmdlet creates a file

JamesWTruher opened the pull request:
Implementation of https://github.com/PowerShell/PowerShell-RFC/blob/master/1-Draft/RFC0020-DefaultFileEncoding.md

Cmdlets now create files with consistent encoding (UTF8 without BOM) on all platforms.
A new preference variable `PSDefaultFileEncoding` is now available to enable users to set encoding for cmdlets. By setting `$PSDefaultFileEncoding = "WindowsLegacy"` users can select the encoding which exists in PowerShell5 for each specific cmdlet.

Provider and cmdlet encoding methods have been centralized and are now common.

JamesWTruher committed:
Unify file encoding and set new default to UTF8 without BOM

Create new class PowerShellEncoding and enum FileEncoding to unify cmdlet and provider code for file encoding.
Created PowerShellEncoding class and FileEncoding enum and removed ClrFacade.GetDefaultEncoding.
PSDefaultFileEncoding preference variable now can set file encoding across all cmdlets.
Setting PSDefaultFileEncoding to WindowsLegacy will set file encoding to historic PowerShell5 encodings.

JamesWTruher committed:
Fix test issue where newline was being provided rather than calculated

some tests were failing on Windows because new line is different, calculate the bytes
in newline rather than hardcoding them

JamesWTruher committed:
Remove code which was relying on previous encoding code

ClrFacade retains some of its functionality, but now relies on PowerShellEncoding class for
knowing what the default coding is. The encoding methods which calls native methods is retained.

JamesWTruher committed:
fix another couple of xplat issues

the WindowsLegacy behavior for New-ModuleManifest should get the correct number of
bytes which will change depending on how many bytes are encoded for [Environment]::NewLine

iSazonov commented on src/System.Management.Automation/utils/Encoding.cs:
```diff
@@ -0,0 +1,387 @@
+        /// </summary>
+        public static Encoding GetDefaultEncoding()
+        {
+            return new UTF8Encoding(false);
```
We can use static cache for `UTF8Encoding(false);` here and in Line 142 too - the cache is in Line 311.

JamesWTruher replied:
good point

lzybkr commented:
Please fix the conflict and rebase.

JamesWTruher committed:
Return static Utf8NoBom encoder rather than creating a new instance

update calls which create had been creating a new instance of the Utf8 encoding without BOM
to return the available static

JamesWTruher committed:
Change expected encoding to not distinguish based on platform except for newline

iSazonov commented:
If the new test batch covers all the old encoding variants, it would be good to merge the tests before the PR - then we'll be sure to see that the PR has full backward compatibility.

adityapatwardhan commented on src/System.Management.Automation/utils/PathUtils.cs:
```diff
@@ -17,6 +18,42 @@ namespace System.Management.Automation
+        }
+
+        /*
         /// <summary>
```
Delete the commented block?

adityapatwardhan commented on src/System.Management.Automation/utils/PathUtils.cs:
```diff
@@ -188,6 +226,13 @@ internal static void ReportFileOpenFailure(Cmdlet cmdlet, string filePath, Excep
+            return new StreamReader(fileStream, PowerShellEncoding.GetEncoding(command, encoding));
+        }
+
+        /*
```
Delete commented block

JamesWTruher replied:
fixed - I missed that one

adityapatwardhan commented on src/System.Management.Automation/utils/Encoding.cs:
```diff
@@ -0,0 +1,387 @@
+        }
+    }
+
+}
```
Can you check the code coverage with the newly added tests for this file? So we add tests now if we are missing any.

JamesWTruher replied:
now that we can, i will :)

JamesWTruher commented on src/System.Management.Automation/utils/Encoding.cs:
```diff
@@ -0,0 +1,387 @@
+                { "microsoft.powershell.commands.setcontentcommand", Encoding.ASCII },
+                // Providers are handled here
+                { "microsoft.powershell.commands.filesystemprovider", Encoding.ASCII },
+
```
fixed

JamesWTruher commented on src/System.Management.Automation/utils/Encoding.cs:
```diff
@@ -0,0 +1,387 @@
+
+        internal static Encoding GetWindowsLegacyEncoding(string name)
+        {
+            if ( legacyEncodingMap.ContainsKey(name))
```
ok

JamesWTruher commented on src/System.Management.Automation/utils/Encoding.cs:
```diff
@@ -0,0 +1,387 @@
+        {
+            Encoding resolvedEncoding = GetDefaultEncoding();
+            FileEncoding encodingPreference = GetEncodingPreference(provider.SessionState);
+            if ( encoding == FileEncoding.Unknown && encodingPreference != FileEncoding.Unknown )
```
yes - because of the provider cmdlets which leave the encoding to the underlying provider as a dynamic parameter it needs to be handled via a different path than checking against the name of the cmdlet.

dantraMSFT commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/CSVCommands.cs:
```diff
@@ -206,8 +207,7 @@ public SwitchParameter NoClobber
         [Parameter()]
-        [ValidateSetAttribute(new string[] { "Unicode", "UTF7", "UTF8", "ASCII", "UTF32", "BigEndianUnicode", "Default", "OEM" })]
-        public string Encoding { get; set; }
+        public FileEncoding Encoding { get; set; } = FileEncoding.Unknown;
```
Why aren't you using a ValidateSet with the FileEncoding values?

iSazonov replied:
We could use #3784 - dynamic validate set, don't expose `FileEncoding` and  allow enhancement to use legacy charsets.

JamesWTruher replied:
I don't want to use validateset here at all. Especially since this is so pervasive in so many areas, I get the same more easily with an enum

dantraMSFT commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/CSVCommands.cs:
```diff
@@ -571,8 +571,7 @@ public SwitchParameter UseCulture
         [Parameter()]
-        [ValidateSetAttribute(new[] { "Unicode", "UTF7", "UTF8", "ASCII", "UTF32", "BigEndianUnicode", "Default", "OEM" })]
-        public string Encoding { get; set; }
+        public FileEncoding Encoding { get; set; } = FileEncoding.Unknown;
```
As above, Why aren't you using a ValidateSet with the FileEncoding values

JamesWTruher replied:
same as above

dantraMSFT commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/MatchString.cs:
```diff
@@ -1200,17 +1200,7 @@ public SwitchParameter AllMatches
-            EncodingConversion.Default,
-            EncodingConversion.OEM })]
-        public string Encoding { get; set; }
+        public FileEncoding Encoding { get; set; } = FileEncoding.Unknown;
```
The default should be FileEncoding.UTF8. See BeginProcessing.

JamesWTruher replied:
i want to be sure that I discover the encoding. If I set it to something that is likely to be provided, I won't be able to determine whether the user explicitly set it. By setting it to unknown, I'll calculate the appropriate value and I can tell that it wasn't set by the user

dantraMSFT commented on src/System.Management.Automation/utils/Encoding.cs:
```diff
@@ -0,0 +1,387 @@
+        /// </summary>
+        public static Encoding GetProviderEncoding(CmdletProvider provider, FileEncoding encoding)
+        {
+            Encoding resolvedEncoding = GetDefaultEncoding();
```
Suggest checking for encoding != FileEncoding.Unknown first; none of the other code paths are executed for that case.

dantraMSFT commented on src/System.Management.Automation/utils/Encoding.cs:
```diff
@@ -0,0 +1,387 @@
+                // the parameter is not specifically set, so check the preference variable
+                encodingPreference = GetEncodingPreference(cmdlet.Context.SessionState);
+                // If set to unknown, we accept that it is unset
+                preferenceSetAndValid = encodingPreference != FileEncoding.Unknown; 
```
dead code?

dantraMSFT replied:
Dead code.

dantraMSFT commented on src/System.Management.Automation/utils/Encoding.cs:
```diff
@@ -0,0 +1,387 @@
+        {
+            Encoding resolvedEncoding = GetDefaultEncoding();
+            FileEncoding encodingPreference = FileEncoding.Unknown;
+            bool preferenceSetAndValid = false;
```
dead code?

dantraMSFT replied:
Dead code.

JamesWTruher replied:
removed

dantraMSFT commented on src/System.Management.Automation/utils/Encoding.cs:
```diff
@@ -0,0 +1,387 @@
+                {
+                    using (BinaryReader reader = new BinaryReader(stream))
+                    {
+                        bytesRead = reader.Read(initialBytes, 0, 100);
```
use initialBytes.Length instead of 100.

JamesWTruher replied:
fixed

dantraMSFT commented on src/System.Management.Automation/utils/Encoding.cs:
```diff
@@ -0,0 +1,387 @@
+                return FileEncoding.Default;
+            }
+
+            byte[] initialBytes = new byte[100];
```
Do you really need to read 100 bytes?

JamesWTruher replied:
this was part of a refactor where this code has been around for a long time, and I didn't really want to change it. I'm certainly open reducing the size, but it's not too much, not too little and any number here would be a guess anyway, right?

dantraMSFT commented on src/System.Management.Automation/utils/Encoding.cs:
```diff
@@ -0,0 +1,387 @@
+        internal static readonly UTF8Encoding utf8NoBom = new UTF8Encoding(encoderShouldEmitUTF8Identifier: false);
+
+        // take a look at the file contents and guess at the best encoding
+        internal static FileEncoding GetEncoding(string path)
```
Suggest naming this GetFileEncoding for clarity.

JamesWTruher replied:
fixed

dantraMSFT commented on src/System.Management.Automation/utils/Encoding.cs:
```diff
@@ -0,0 +1,387 @@
+        {
+            if (!File.Exists(path))
+            {
+                return FileEncoding.Default;
```
I think this should probably throw; not return Default.

JamesWTruher replied:
not necessarily - the file could be later created, so I need to pass back something - this code was refactored from another location, so i wanted to do as little as possible

dantraMSFT commented on src/System.Management.Automation/utils/Encoding.cs:
```diff
@@ -0,0 +1,387 @@
+
+            if (bytesRead > 3)
+            {
+                preamble = String.Join("-", initialBytes[0], initialBytes[1], initialBytes[2], initialBytes[3]);
```
I suspect the string building/lookups appears overkill.  Consider testing the bytes directly?

JamesWTruher replied:
this was code which was refactored from another location, I didn't really want to change it

dantraMSFT commented on src/System.Management.Automation/commands/utility/FormatAndOutput/common/FormatXMLWriter.cs:
```diff
@@ -37,7 +38,7 @@ internal class FormatXmlWriter
             FileStream fileStream;
             FileInfo fileInfo;
-            PathUtils.MasterStreamOpen(cmdlet, filepath, "ascii", true, false, force, noclobber,
+            PathUtils.MasterStreamOpen(cmdlet, filepath, PowerShellEncoding.GetEncoding(FileEncoding.Ascii), true, false, force, noclobber,
```
This appears to be the only external call to Encoding overload of PathUtils.MasterStreamOpen.  Suggest using FileEncoding.Ascii here and merge the two MasterStreamOpen overloads.

JamesWTruher replied:
yah - I was worried about removing the overload as it was a public interface and thought to do less harm, just in case someone else was using it somehow.

dantraMSFT replied:
MasterStreamOpen is internal.

JamesWTruher replied:
I'll investigate

JamesWTruher replied:
the encoding overload is used by CSV cmdlets which do their own determination of the encoding (in the append case), and pass that along to the open. I'd rather not refactor all that code if I don't need to.

dantraMSFT commented on src/System.Management.Automation/engine/InitialSessionState.cs:
```diff
@@ -4824,6 +4825,7 @@ .ForwardHelpCategory Cmdlet
         internal const ActionPreference defaultVerbosePreference = ActionPreference.SilentlyContinue;
         internal const ActionPreference defaultWarningPreference = ActionPreference.Continue;
         internal const ActionPreference defaultInformationPreference = ActionPreference.SilentlyContinue;
+        internal const Microsoft.PowerShell.FileEncoding defaultFileEncodingPreference = FileEncoding.Unknown;
```
I can find no references to this field other than setting the session state variable. the variable. How is it used and how does it control behavior? Currently, it appears to be a NOP.

JamesWTruher replied:
it's used down in line 4915 where it essentially lays claim to the variable

dantraMSFT commented on src/System.Management.Automation/engine/Utils.cs:
```diff
@@ -8,6 +8,7 @@
 using System.Management.Automation.Internal;
 using System.Management.Automation.Security;
 using System.Reflection;
+using Microsoft.PowerShell;
```
Is this actually needed? The only other change to this file is the removal of the GetEncoding logic.

JamesWTruher replied:
removed

dantraMSFT commented on src/System.Management.Automation/engine/Modules/ScriptAnalysis.cs:
```diff
@@ -95,7 +96,7 @@ internal static string ReadScript(string path)
             using (FileStream readerStream = new FileStream(path, FileMode.Open, FileAccess.Read))
             {
-                Encoding defaultEncoding = ClrFacade.GetDefaultEncoding();
+                Encoding defaultEncoding = PowerShellEncoding.GetDefaultEncoding();
```
Shouldn't this be using PowerShellEncoding.GetEncoding

dantraMSFT commented on src/System.Management.Automation/utils/ClrFacade.cs:
```diff
@@ -209,7 +210,7 @@ internal static Encoding GetOEMEncoding()
 #if UNIX        // PowerShell Core on Unix
-                s_oemEncoding = GetDefaultEncoding();
+                s_oemEncoding = PowerShellEncoding.GetDefaultEncoding();
 #elif CORECLR   // PowerShell Core on Windows
```
I think the CORECLR branches could use commenting in both GetDefaultEncoding and GetOEMEncoding. The logic reasoning for GetACP and GetOEMCP is opaque.

iSazonov replied:
The discussion was in #3248 and fixed in #3467. Maybe it should be reviewed again. In short we returned behavior Windows PowerShell.

JamesWTruher replied:
I'll add comments about this, we can change it later after we have live-fire feedback.

dantraMSFT commented on src/System.Management.Automation/utils/ClrFacade.cs:
```diff
@@ -355,6 +356,7 @@ private static SecurityZone ReadFromZoneIdentifierDataStream(string filePath)
                                             FileAccess.Read, FileShare.Read);
 
                 // If we successfully get the zone data stream, try to read the ZoneId information
+                // use the method in this class not PowerShellEncoding.
```
Why  isn't this detecting the encoding of the stream data?

dantraMSFT commented on src/System.Management.Automation/utils/PathUtils.cs:
```diff
@@ -17,6 +18,42 @@ namespace System.Management.Automation
+        /// <param name="streamWriter">Result2: <see cref="StreamWriter"/> (inherits from <see cref="TextWriter"/>) opened for writing</param>
+        /// <param name="readOnlyFileInfo">Result3: file info that should be used to restore file attributes after done with the file (<c>null</c> is this is not needed)</param>
+        /// <param name="isLiteralPath">True if wildcard expansion should be bypassed.</param>
+        internal static void MasterStreamOpen(
```
Suggest merging this with the Encoding overload. There appears to be only 1 caller to the later overload and it starts converts a FileEncoding to an Encoding to accomplish it.

dantraMSFT commented on src/System.Management.Automation/utils/PathUtils.cs:
```diff
@@ -490,7 +537,7 @@ internal static Encoding Convert(Cmdlet cmdlet, string encoding)
 
             if (string.Equals(encoding, Default, StringComparison.OrdinalIgnoreCase))
-                return ClrFacade.GetDefaultEncoding();
+                return PowerShellEncoding.GetDefaultEncoding();
```
Suggest using a dictionary for these cascading 'if' tests.

JamesWTruher committed:
Remove commented out code

dantraMSFT commented on test/powershell/engine/Encoding.Tests.ps1:
```diff
@@ -0,0 +1,173 @@
+            $encoder = [Microsoft.PowerShell.PowerShellEncoding]::GetEncoding($encoding)
+            $encoder.GetBytes([Environment]::NewLine) -Join "-"
+        }
+        function Test-GetEncoding
```
This function is not referenced.

JamesWTruher replied:
removed

JamesWTruher replied:
removed

dantraMSFT commented on test/powershell/engine/Encoding.Tests.ps1:
```diff
@@ -0,0 +1,173 @@
+        }
+
+        $preambleTests =
+            @{ Name = 'Ascii'; Preamble = '' },
```
The values that have an empty Preamble mask possible bugs and unexpected preamble may be present but it won't be detected resulting in a false positive. For example, the test won't detect that a UTF32 encoding was encountered when ASCII was expected.
I suggest coding the test data to have a well known sequence following the BOM (or lack of BOM) and have the Preamble strings include this sequence. This will help reduce potential false positives.

dantraMSFT commented on test/powershell/engine/Encoding.Tests.ps1:
```diff
@@ -0,0 +1,173 @@
+            @{ Name = 'WindowsLegacy'; Preamble = '' }
+
+        $contentTests =
+            @{ Name = 'Ascii'; Bytes = "116-63-115-116-" + (Get-NewLineBytes Ascii) },
```
Minor nit; I suggest using Encoding instead of Name.

JamesWTruher replied:
will change

dantraMSFT commented on test/powershell/engine/Encoding.Tests.ps1:
```diff
@@ -0,0 +1,173 @@
+        param ( $Name, $Bytes )
+        $PSDefaultFileEncoding = $Name
+        $testString | out-file -encoding ascii $testFile
+        Get-FileBytes $testFile | should be ("116-63-115-116-" + (Get-NewLineBytes ASCII))
```
A brief comment explaining why this is handled specially with the leading bytes spelled out would be useful for this test and subsequent tests.

JamesWTruher replied:
will add

JamesWTruher replied:
i decided to refactor these so I could avoid hardcoding these strings more than once

JamesWTruher committed:
refactor tests to use fewer instances of hardcoded byte strings

only use hardcoded bytes when it's a custom file generation
(like Export-CliXml or New-ModuleManifest)
or when we're looking at a set of partial results
also remove an unused function

JamesWTruher committed:
update tests to include tests against Encoding probe method for existing files

improve code coverage for PowerShellEncoding class and don't duplicate byte representations when they're not needed

JamesWTruher committed:
Refactor Encoding.cs to improve readability

Remove a couple of extraneous using statements
Move encoding code from PathUtils.cs to Encoding.cs
Move and rename GetEncoding method in Utils.cs to GetFileEncodingFromFile in Encoding.cs
Expand some explanatory comments with more details

iSazonov commented on src/System.Management.Automation/utils/Encoding.cs:
```diff
@@ -381,6 +298,101 @@ internal static FileEncoding GetEncoding(string path)
+        /// Get the Windows legacy encoding from our encoding map
+        internal static Encoding GetWindowsLegacyEncoding(string name)
+        {
+            if ( legacyEncodingMap.ContainsKey(name))
```
Can we use `return TryGetValue(name, out Encoding value) ? value : Encoding.Default`?

iSazonov replied:
Ah, we can use `GetValueOrDefault`
```c#
return legacyEncodingMap.GetValueOrDefault(name, Encoding.Default);
```

iSazonov commented on src/System.Management.Automation/utils/Encoding.cs:
```diff
@@ -151,7 +152,9 @@ public static Encoding GetEncoding(FileEncoding TextEncoding)
-                    result = Encoding.BigEndianUnicode;
+                    // This can possibly throw, but if so, we can't provide
+                    // the encoding which the user requested, so we should fail
+                    result = Encoding.GetEncoding("utf-32BE");
```
Should we cache this in  internal static readonly constant as for `internal static readonly UTF8Encoding utf8NoBom`?

iSazonov commented on src/System.Management.Automation/utils/Encoding.cs:
```diff
@@ -381,6 +298,101 @@ internal static FileEncoding GetEncoding(string path)
+            FileEncoding encodingPreference = FileEncoding.Unknown;
+            try
+            {
+                // It doesn't matter if this fails or throws, we will return unknown in that case
```
Please replace `unknown` on 'FileEncoding.Unknown'.

JamesWTruher commented:
@adityapatwardhan code coverage with latest tests is about 98%

adityapatwardhan reviewed (approved):
LGTM

lzybkr commented on test/powershell/engine/Encoding.Tests.ps1:
```diff
@@ -0,0 +1,225 @@
+    }
+
+    It "getting the encoding for an unknown cmdlet should return utf-8" {
+        $method = [microsoft.powershell.powershellencoding].getmember("GetWindowsLegacyEncoding","NonPublic,Static")
```
We'd like to avoid reflection as much as possible in tests - can you find user accessible way to test this instead, e.g. calling a public api that calls this internal api?

JamesWTruher replied:
this is pretty difficult as the API that is public takes a cmdlet object, so i couldn't think of any other way to do this. I'm happy to hear suggestions

lzybkr replied:
What's wrong with `[cmdlet]::new()` to get that instance?

lzybkr commented on src/System.Management.Automation/utils/Encoding.cs:
```diff
@@ -0,0 +1,399 @@
+            try
+            {
+                // It doesn't matter if this fails or throws, we will return unknown in that case
+                object tmp = sessionState.PSVariable.GetValue("PSDefaultFileEncoding");
```
You should probably also test `sessionState != null` to avoid the expensive exception because I can see this method getting called with a null session state.

lzybkr commented on src/System.Management.Automation/engine/SpecialVariables.cs:
```diff
@@ -289,6 +293,7 @@ internal static class SpecialVariables
                                                                     /* WarningPreference */     typeof(ActionPreference),
                                                                     /* InformationPreference */ typeof(ActionPreference),
                                                                     /* ConfirmPreference */     typeof(ConfirmImpact),
+                                                                    /* PSDefaultFileEncoding */ typeof(Microsoft.PowerShell.FileEncoding),
```
This line should be removed. I realize it's not obvious why this array exists, but it is used for variables that are always created, and these preference variables get created in every script cmdlet, but your new preference variable should not.

JamesWTruher replied:
fixed

lzybkr commented on src/System.Management.Automation/namespaces/FileSystemContentStream.cs:
```diff
@@ -1160,8 +1161,10 @@ internal FileStreamBackReader(FileStream fileStream, Encoding encoding)
                 // Get the oem encoding and system current ANSI code page
-                _oemEncoding = EncodingConversion.Convert(null, EncodingConversion.OEM);
-                _defaultAnsiEncoding = EncodingConversion.Convert(null, EncodingConversion.Default);
+                // _oemEncoding = EncodingConversion.Convert(null, EncodingConversion.OEM);
```
Delete?

JamesWTruher replied:
fixed

lzybkr reviewed (changes requested):
The @PowerShell/powershell-committee should probably review the new public apis.

JamesWTruher commented on src/System.Management.Automation/utils/Encoding.cs:
```diff
@@ -0,0 +1,399 @@
+    /// <summary>
+    /// the helper class for determining encodings for PowerShell
+    /// </summary>
+    public static class PowerShellEncoding
```
Please suggest something

lzybkr replied:
EncodingUtils or something like that?

JamesWTruher replied:
changed to EncodingUtils

JamesWTruher commented on src/System.Management.Automation/utils/Encoding.cs:
```diff
@@ -0,0 +1,399 @@
+        /// <param name="TextEncoding">The enum value</param>
+        /// <returns>System.Text.Encoding</returns>
+        /// </summary>
+        public static Encoding GetEncoding(FileEncoding TextEncoding)
```
fixed

JamesWTruher replied:
i see this as an opportunity to return the default encoding (utf8NoBom) - making this an error doesn't seem to be needed

lzybkr replied:
Shouldn't I be able to use this api to get an Encoding for a file I've written using `FileEncoding`? That will work for everything but `WindowsLegacy`, so I would say it's better to throw.

JamesWTruher commented on src/System.Management.Automation/utils/Encoding.cs:
```diff
@@ -0,0 +1,399 @@
+        /// </summary>
+        public static Encoding GetEncoding(FileEncoding TextEncoding)
+        {
+            System.Text.Encoding result = GetDefaultEncoding();
```
fixed

JamesWTruher commented on src/System.Management.Automation/utils/Encoding.cs:
```diff
@@ -0,0 +1,399 @@
+        /// <param name="encoding">The Encoding parameter value</param>
+        /// <returns>System.Text.Encoding</returns>
+        /// </summary>
+        public static Encoding GetEncoding(Cmdlet cmdlet, FileEncoding encoding)
```
i fixed the public api parameter validation but I disagree about the throw - if I don't have access to session state, I can't retrieve the value I can still return that no encoding preference has been set

JamesWTruher commented on src/System.Management.Automation/utils/Encoding.cs:
```diff
@@ -0,0 +1,399 @@
+            Encoding resolvedEncoding = GetDefaultEncoding();
+            FileEncoding encodingPreference = FileEncoding.Unknown;
+            bool preferenceSetAndValid = false;
+            string name = cmdlet.GetType().FullName.ToLower(CultureInfo.InvariantCulture);
```
yah, especially since the dictionary is a case insensitive one

JamesWTruher commented on src/System.Management.Automation/utils/Encoding.cs:
```diff
@@ -0,0 +1,399 @@
+        /// <param name="path">The path to a file to inspect for an encoding</param>
+        /// <returns>System.Text.Encoding</returns>
+        /// </summary>
+        public static FileEncoding GetFileEncodingFromFile(string path)
```
concrete suggestion please? This code has existed for some time (although not public) and has been used for a long time

lzybkr replied:
Don't make it public, or do a ton more testing and fix the issues I raised.

JamesWTruher replied:
changed to internal

JamesWTruher commented on src/System.Management.Automation/utils/Encoding.cs:
```diff
@@ -0,0 +1,399 @@
+
+            // Test for four-byte preambles
+            string preamble = null;
+            FileEncoding foundEncoding = FileEncoding.Default;
```
fixed

JamesWTruher commented on src/System.Management.Automation/utils/Encoding.cs:
```diff
@@ -0,0 +1,399 @@
+        // If the user sets the default encoding to WindowsLegacy, we will
+        // be able to encode for that
+        internal static Dictionary<String, Encoding> legacyEncodingMap =
+            new Dictionary<string, Encoding>(StringComparer.OrdinalIgnoreCase)
```
i didn't want to use that because this means that those types must be known in the engine, which I don't believe is a good thing.

JamesWTruher replied:
also, with regard to the test, pester doesn't do array comparisons very well. This is a reasonable approach for that problem

lzybkr replied:
We don't want to introduce a new circular dependency, but we have implicit knowledge encoded here, either via a string or `typeof` - there is little difference.

Actually removing the implicit knowledge probably isn't worth it, but you could do that via registration - each of these cmdlets would call an internal api to register itself.

JamesWTruher replied:
your example also misses the UTF8 without BOM case and looks return the wrong encoding (ASCII) in the case that it encountered such a file.

lzybkr replied:
Yeah - my sample served it's purpose and nothing more - but I mentioned it to show it's not much code.

JamesWTruher replied:
Text.Encoding.UTF7.GetPreamble() returns no bytes

JamesWTruher commented on test/powershell/Language/Parser/RedirectionOperator.Tests.ps1:
```diff
@@ -29,22 +29,23 @@ Describe "Redirection operator now supports encoding changes" -Tags "CI" {
         $expectedBytes = .{ $BOM; $TXT; $CR }
         $bytes.Count | should be $expectedBytes.count
         for($i = 0; $i -lt $bytes.count; $i++) {
             $bytes[$i] | Should be $expectedBytes[$i]
```
fixed

JamesWTruher commented on src/System.Management.Automation/utils/Encoding.cs:
```diff
@@ -0,0 +1,399 @@
+        /// <summary>
+        /// No encoding, or unset.
+        /// </summary>
+        Unknown,
```
how would this be set back to -1, once set? would the user be stuck with the value?

lzybkr replied:
Internally, we would have a const, similar to what the CLR does for colors (see [here](https://github.com/dotnet/corefx/blob/ef82e74c065a6f839515fee8871e1114e6aac357/src/System.Console/src/System/Console.cs#L170)).

I don't quite follow your question - but in C#, you can assign any value you like (with a cast). In PowerShell it's not as easy because we check an integer value to see if it's a valid enumerator.

We could also consider using the value `0` to mean `Unspecified` as long as the first enum starts at 1 - this has the benefit of being the default value if a field is not assigned.

JamesWTruher committed:
Change FileEncoding.Unknown to FileEncoding.Unspecified which more closer to what is really happening

Update RedirectionOperator tests to compare bytes in a more sensible manner
Remove PSDefaultFileEncoding from special variable collection so they don't show up in script cmdlets
miscellaneous code clean up

JamesWTruher commented on test/powershell/engine/Encoding.Tests.ps1:
```diff
@@ -0,0 +1,225 @@
+            Oem = ''
+            String = '255-254'
+            UTF32 = '255-254-0-0'
+            UTF7 = ''
```
Text.Encoding.UTF7.GetPreamble() returns no bytes

lzybkr replied:
I understand (maybe the api designer didn't think it was worth returning an array of possible preambles, and the implementer decided it was better to return nothing rather than choosing one possible preamble.)

Regardless, it's worth a comment on why utf7 is missing an expected bom, after all, we aren't calling it UTF7NoBOM.

JamesWTruher committed:
Changed PowerShellEncoding class name to EncodingUtils

made file encoding probe method internal

JamesWTruher committed:
update PowerShellEncoding to use new name EncodingUtils

SteveL-MSFT commented:
@PowerShell/powershell-committee reviewed this and we should have an RFC specific to the new public apis (and generally any new public apis should be RFCs going forward)

JamesWTruher committed:
Use new class name EncodingUtils

SteveL-MSFT commented on src/System.Management.Automation/utils/Encoding.cs:
```diff
@@ -0,0 +1,402 @@
+        /// <param name="path">The path to a file to inspect for an encoding</param>
+        /// <returns>System.Text.Encoding</returns>
+        /// </summary>
+        internal static FileEncoding GetFileEncodingFromFile(string path)
```
Seems like this should be `GetEncodingFromFile` to keep the naming consistent

lzybkr commented on src/System.Management.Automation/utils/Encoding.cs:
```diff
@@ -0,0 +1,402 @@
+            FileEncoding encodingPreference = FileEncoding.Unspecified;
+            try
+            {
+                // It doesn't matter if this fails or throws, we will return unknown in that case
```
return Unspecified, not return unknown

lzybkr commented on src/System.Management.Automation/utils/Encoding.cs:
```diff
@@ -0,0 +1,402 @@
+        /// <param name="encoding">The Encoding parameter value</param>
+        /// <returns>System.Text.Encoding</returns>
+        /// </summary>
+        public static Encoding GetEncoding(Cmdlet cmdlet, FileEncoding encoding)
```
Maybe this api would be better if we passed a `SessionState` instead of `Cmdlet`.

mklement0 commented on src/System.Management.Automation/utils/Encoding.cs:
```diff
@@ -0,0 +1,402 @@
+                { "239-187-191", FileEncoding.UTF8BOM },
+            };
+
+        internal static char[] nonPrintableCharacters = {
```
Strictly speaking, `11` (vertical tab) and `12` (form feed) are also printable characters; given their rarity, I assume they were left out intentionally; if so, perhaps a comment would appease drive-by pedants like me.

mklement0 commented on src/System.Management.Automation/utils/ClrFacade.cs:
```diff
@@ -126,15 +127,20 @@ internal static Encoding GetDefaultEncoding()
+#else           // PowerShell Core on Windows, which needs provider registration
                 EncodingRegisterProvider();
 
                 uint oemCp = NativeMethods.GetOEMCP();
```
Looks like this line is no longer needed.

mklement0 commented on src/System.Management.Automation/utils/ClrFacade.cs:
```diff
@@ -126,15 +127,20 @@ internal static Encoding GetDefaultEncoding()
+            // The OEM code pages are sometimes used by Win32 console applications, and
+            // on non-Windows platforms they still may have uses (if installed) and
+            // could be used if desired.
+            // On non-windows platforms, they have more limited uses, and probably won't
```
This sentence is a bit confusing. Doesn't the previous one say everything that needs to be said?

mklement0 commented on src/System.Management.Automation/utils/Encoding.cs:
```diff
@@ -0,0 +1,402 @@
+                { "microsoft.powershell.commands.setcontentcommand", Encoding.ASCII },
+                // Providers are handled here
+                { "microsoft.powershell.commands.filesystemprovider", Encoding.ASCII },
+            };
```
Unfortunately, that doesn't appear to be correct.

As for the _writing_ cmdlets:

`Add-Content` and `Set-Content` - despite what the documentation claims - have always used "ANSI" encoding, not ASCII - see https://github.com/PowerShell/PowerShell-Docs/issues/1483.

What about `Send-MailMessage`? The help topics that ASCII encoding is the default.

`New-Item -Type File -Value` currently creates BOM-less(!) UTF-8.

As an aside: While `Export-Csv` without `-Append` indeed creates ASCII files, `Export-Csv -Append` currently blindly appends _UTF-8_ if the existing file's encoding is any of ASCII/UTF-8/"ANSI", but correctly matches UTF-16LE and UTF-16BE.

---

As for the _reading_ cmdlets:

`Get-Content` currently defaults to "ANSI" in the absence of a BOM, as does `Import-PowerShellDataFile`.

`Import-Csv` currently actually assumes UTF-8 in the absence of a BOM, as do `Import-CliXml` and `Select-String`

JamesWTruher commented:
i am going to take a different route for this change - closing this
